### PR TITLE
[Merged by Bors] - feat: weaken assumptions for `IsCompact.existsIsLeast` and all of its variations

### PR DESCRIPTION
**Pull request by ADedecker**
As discussed a while ago on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/upper.20semicontinuous.20functions.20are.20bounded.20below.20compact), we introduce classes expressing that a certain topology has closed `Ici`s/`Iic`s, which is sufficient to get boundedness of compacts on the desired side. The main application is that these now apply to types satisfying [UpperTopology](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Order/LowerUpperTopology.html#UpperTopology)/[LowerTopology](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Order/LowerUpperTopology.html#LowerTopology), which will allow us to apply these compactness results to semicontinuous functions.

The naming was discussed [here](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/upper.20semicontinuous.20functions.20are.20bounded.20below.20compact/near/383557463)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Order/Basic.lean [hunk 1]
```diff
@@ -86,6 +86,20 @@ universe u v w
...
+export LowerClosedTopology (isClosed_le')
+#align is_closed_le' LowerClosedTopology.isClosed_le'
+
+class UpperClosedTopology (α : Type _) [TopologicalSpace α] [Preorder α] : Prop where
+  isClosed_ge' (a : α) : IsClosed { b : α | a ≤ b }
+
+export LowerClosedTopology (isClosed_le')
+#align is_closed_ge' UpperClosedTopology.isClosed_ge'
+
+export UpperClosedTopology (isClosed_ge')
```
Comment (j-loreaux) [inline Mathlib/Topology/Order/Basic.lean:101]: I think you mean:
```suggestion
export UpperClosedTopology (isClosed_ge')
#align is_closed_ge' UpperClosedTopology.isClosed_ge'
```

File: Mathlib/Topology/Algebra/Order/Compact.lean [hunk 2]
```diff
@@ -378,61 +382,89 @@ end ConditionallyCompleteLinearOrder
...
-    ∀ {s : Set β},
-      IsCompact s → s.Nonempty → ∀ {f : β → α}, ContinuousOn f s → ∃ x ∈ s, sSup (f '' s) = f x :=
-  @IsCompact.exists_sInf_image_eq αᵒᵈ _ _ _ _ _
+theorem IsCompact.exists_sSup_image_eq [UpperClosedTopology α] {s : Set β} (hs : IsCompact s)
+    (ne_s : s.Nonempty) : ∀ {f : β → α}, ContinuousOn f s → ∃ x ∈ s, sSup (f '' s) = f x :=
+  IsCompact.exists_sInf_image_eq (α := αᵒᵈ) hs ne_s
 #align is_compact.exists_Sup_image_eq IsCompact.exists_sSup_image_eq
 
+-- porting note: todo: add dual versions
+
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/Order/Compact.lean:444]: I know you didn't write this, but do you want to address it now while we're at it? If you want to put it in a dependent PR that's fine.

File: Mathlib/Topology/Algebra/Order/Compact.lean [hunk 3]
```diff
@@ -378,61 +382,89 @@ end ConditionallyCompleteLinearOrder
...
+theorem IsCompact.exists_isMinOn_mem_subset [LowerClosedTopology α] {f : β → α} {s t : Set β}
+    {z : β} (ht : IsCompact t) (hf : ContinuousOn f t) (hz : z ∈ t)
+    (hfz : ∀ z' ∈ t \ s, f z < f z') : ∃ x ∈ s, IsMinOn f t x :=
+  let ⟨x, hxt, hfx⟩ := ht.exists_isMinOn ⟨z, hz⟩ hf
+  ⟨x, by_contra <| fun hxs => (hfz x ⟨hxt, hxs⟩).not_le (hfx hz), hfx⟩
+
+@[deprecated IsCompact.exists_isMinOn_mem_subset]
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/Order/Compact.lean:451]: why is this deprecated? do you know?

File: Mathlib/Topology/Order/LowerUpperTopology.lean [hunk 4]
```diff
@@ -273,15 +273,17 @@ instance instUpperTopologyDual [Preorder α] [TopologicalSpace α] [LowerTopolog
     UpperTopology (αᵒᵈ) where
   topology_eq_upperTopology := topology_eq_lowerTopology (α := α)
 
-/-- Left-closed right-infinite intervals [a, ∞) are closed in the lower topology. -/
```
Comment (j-loreaux) [inline Mathlib/Topology/Order/LowerUpperTopology.lean:276]: Let's leave the docstring on this instance.

File: Mathlib/Topology/Order/LowerUpperTopology.lean [hunk 5]
```diff
@@ -380,9 +382,8 @@ instance instLowerTopologyDual [Preorder α] [TopologicalSpace α] [UpperTopolog
     LowerTopology (αᵒᵈ) where
   topology_eq_lowerTopology := topology_eq_upperTopology (α := α)
 
-/-- Left-infinite right-closed intervals (-∞,a] are closed in the upper topology. -/
```
Comment (j-loreaux) [inline Mathlib/Topology/Order/LowerUpperTopology.lean:383]: ditto

**Review by j-loreaux — COMMENTED** [review]
Other than naming and the comments below, I think this looks good. I think the current naming is awful because it means the `LowerTopology` is an `UpperClosedTopology`. Personally, I think `ClosedIciTopolology` (or `IciClosedTopology`) would be good candidates, but we should probably have a Zulip thread to solicit recommendations and vote.

File: Mathlib/Topology/Algebra/Order/Compact.lean (hunk 2 above)
Comment (ADedecker) [reply Mathlib/Topology/Algebra/Order/Compact.lean]: Done, except for the deprecated one.

**Comment by ADedecker** [comment]
I asked for the name on Zulip, otherwise this is ready for review again

**Comment by jcommelin** [comment]
What is the outcome of the zulip poll? Could you please link to the poll from this PR?

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by ADedecker** [comment]
> What is the outcome of the zulip poll? Could you please link to the poll from this PR?

There wasn't a lot of votes, but they are in favor of the current naming. I've linked the poll in the PR description.

**Comment by ADedecker** [comment]
bors cancel

I will fix it tomorrow.

**Pull request closed without merging**